src/ServerRequestFactoryInterface.php

```diff
@@ -12,20 +12,9 @@
      *
      * @param string $method
      * @param UriInterface|string $uri
+     * @param array $server optional server params ($_SERVER or similar structure)
      *
      * @return ServerRequestInterface
      */
-    public function createServerRequest($method, $uri);
-
-    /**
-     * Create a new server request from server variables.
-     *
-     * @param array $server Typically $_SERVER or similar structure.
-     *
-     * @return ServerRequestInterface
-     *
-     * @throws \InvalidArgumentException
-     *  If no valid method or URI can be determined.
-     */
-    public function createServerRequestFromArray(array $server);
+    public function createServerRequest($method, $uri, $server = []);
```

Comment:
> The $server is not just "server params" but the entire $_SERVER

@shadowhand for a factory that constructs a server-request *from globals*, yes - I don't think we need that, at all.

> Every PSR-7 implementation that I know of includes a method that uses (global) data to create a server request that represents that data

Yes, and if we had the server-request factory that constructs a server-request instance *not* from globals but from the data *as given*, we would need only a single implementation of the factory-class that constructs a server-request from globals.

I'm going to refer to this as the "vanilla" server-request factory.

The vanilla interface is all we need for interop, because, with a single implementation of the global server-request factory, if that implementation depends on an implementation of the vanilla factory, which *doesn't* do any processing of the server-params, this functionality can be implemented once for the entire community - it'll work with any implementation of the vanilla factory.

Figuring out the URI and method, for example, can happen in the server-request from globals factory - the vanilla factory itself will be very easy to implement, it only has to accept the values as given and inject them through a proprietary constructor.

@oscarotero summarized it exactly in his previous and last post.

> In addition to this, we could consider to create a new `ServerRequestFromGlobalsFactoryInterface`

We might consider that, but it's not actually essential in terms of interop - any given project or framework is going to need *one* implementation of a server-request from globals factory, coupled with a vanilla server-request factory implementation. It most likely doesn't matter at all what the method-signature of the global server-request factory looks like, because (assuming it works) there'd be zero reason to replace it.

We could very quickly port Diactoros' global server-request factory to one that uses a vanilla server-request factory instead of Diactoros' own implementation. Like so:

https://gist.github.com/mindplay-dk/b8b779d687dd62fb514183ddfedd2d85/revisions

(quick hack, totally untested.)